tool/net/redbean.c

```diff
@@ -4163,6 +4163,117 @@ static int LuaSetHeader(lua_State *L) {
   return 0;
 }
 
+static void GrowCookieBuffer(char *buf, char *p, size_t *n, size_t increase) {
```

Comment:
The buffering logic here is a little difficult to review.

Some background: I used stpcpy() heavily for the pure C code serving paths because they helped a lot with micro-optimizing every last cycle of performance, so we could hit a million qps. Something like a Lua API has some inherent overhead so it probably isn't worth living as dangerously with the micro-optimizations.

I would be thrilled if you used the new append*() library for this. The functions are all in https://github.com/jart/cosmopolitan/tree/master/libc/stdio They're extremely fast (not quite as fast as stpcpy), well-tested, and well-documented. I'm very proud of them. It took me two years to come up with an API this good. Since if you dig deeper you might notice that there's a whole bunch of other helpers I designed along the way, which turned out to be failures, and are still lingering around as technical debt waiting to be deleted. This new API is safe, easy to review, and easy to use. For example:

```c
char *b=0;
appendf(&b, "%.*s=%.*s", keylen, key, vallen, val);
free(b);
```

Or you could make it 2x faster:

```c
char *b=0;
appendd(&b, key, keylen);
appendw(&b, '=');
appendd(&b, val, vallen);
free(b);
```

Reply:
Ha, I saw this in several places, but went with the same logic you have in SetHeader for consistency (this is where the buffering logic is from). After looking at this in more detail, I love this `append*` API! I'll update the code to use it.

Reply:
😃

Reply:
Updated; I agree, the code is shorter and easier to read.